timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/UseCaseAwareBatchPaxosLearnerResource.java

```diff
@@ -0,0 +1,100 @@
+import com.palantir.paxos.PaxosValue;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("/" + PaxosTimeLockConstants.INTERNAL_NAMESPACE
+        + "/{useCase}/"
+        + PaxosTimeLockConstants.BATCH_INTERNAL_NAMESPACE
+        + "/learner")
+public final class UseCaseAwareBatchPaxosLearnerResource {
+    private final Function<PaxosUseCase, BatchPaxosLearnerResource> delegate;
+
+    public UseCaseAwareBatchPaxosLearnerResource(Function<PaxosUseCase, BatchPaxosLearnerResource> delegate) {
+        this.delegate = delegate;
+    }
+
+    @POST
+    @Path("learn")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Handle(
+            method = HttpMethod.POST,
+            path = "/" + PaxosTimeLockConstants.INTERNAL_NAMESPACE + "/{useCase}/"
+                    + PaxosTimeLockConstants.BATCH_INTERNAL_NAMESPACE + "/learner/learn")
+    public void learn(
+            @PathParam("useCase") @Handle.PathParam(decoder = PaxosUseCaseDecoder.class) PaxosUseCase useCase,
+            @Handle.Body SetMultimap<Client, PaxosValue> paxosValuesByClient) {
+        getBatchLearner(useCase).learn(paxosValuesByClient);
+    }
+
+    @POST
+    @Path("learned-values")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Handle(
+            method = HttpMethod.POST,
+            path = "/" + PaxosTimeLockConstants.INTERNAL_NAMESPACE + "/{useCase}/"
+                    + PaxosTimeLockConstants.BATCH_INTERNAL_NAMESPACE + "/learner/learned-values")
+    public SetMultimap<Client, PaxosValue> getLearnedValues(
+            @PathParam("useCase") @Handle.PathParam(decoder = PaxosUseCaseDecoder.class) PaxosUseCase useCase,
+            @Handle.Body Set<WithSeq<Client>> clientAndSeqs) {
+        return getBatchLearner(useCase).getLearnedValues(clientAndSeqs);
+    }
+
+    @POST
+    @Path("learned-values-since")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Handle(
+            method = HttpMethod.POST,
+            path = "/" + PaxosTimeLockConstants.INTERNAL_NAMESPACE + "/{useCase}/"
+                    + PaxosTimeLockConstants.BATCH_INTERNAL_NAMESPACE + "/learner/learned-values-since")
+    public SetMultimap<Client, PaxosValue> getLearnedValuesSince(
+            @PathParam("useCase") @Handle.PathParam(decoder = PaxosUseCaseDecoder.class) PaxosUseCase useCase,
+            @Handle.Body Map<Client, Long> seqLowerBoundsByClient) {
+        return getBatchLearner(useCase).getLearnedValuesSince(seqLowerBoundsByClient);
+    }
+
+    private BatchPaxosLearnerResource getBatchLearner(PaxosUseCase useCase) {
+        return delegate.apply(useCase);
+    }
+
+    enum PaxosUseCaseDecoder implements ParamDecoder<PaxosUseCase> {
```

Comment:
This is required as Jersey will use PaxosUseCase#fromString to deserialize.

Underdow/Conjure-java will look for valueOf (or a constructor taking a string). Enums have a valueOf that you can't change (and it does not match what we want here).

Depending on observed behavior (did not check yet) around what Conjure-java tries first (constructor or valueOf) when deserializing would allow me to just add a constructor and have it work. That seems fragile.

Custom decoding is explicit, I also intentionally duplicated it in both classes. It's highly couple to resource mounting in Conjure-java.

Additionally, it cannot be made private as bits on resource classes are references in code-gen (separately, endpoint methods for Jersey I suspect have to be public because of runtime reflection).